src/iterator.c

```diff
@@ -1393,61 +1424,68 @@ static int filesystem_iterator_frame_push(
 		path += iter->root_len;
 		path_len -= iter->root_len;
 
+		if (filesystem_iterator_is_dot_git(iter, path, path_len))
+			continue;
+
+		memset(&statbuf, 0, sizeof(statbuf));
+
 		/* examine start / end and the pathlist to see if this path is in it.
 		 * note that since we haven't yet stat'ed the path, we cannot know
 		 * whether it's a directory yet or not, so this can give us an
 		 * expected type (S_IFDIR or S_IFREG) that we should examine)
 		 */
-		if (!filesystem_iterator_examine_path(&dir_expected, &pathlist_match,
+		if (!filesystem_iterator_examine_path(&statbuf, &dir_expected, &diriter, &pathlist_match,
 			iter, frame_entry, path, path_len))
 			continue;
 
 		/* TODO: don't need to stat if assume unchanged for this path and
 		 * we have an index, we can just copy the data out of it.
 		 */
 
-		if ((error = git_path_diriter_stat(&statbuf, &diriter)) < 0) {
-			/* file was removed between readdir and lstat */
-			if (error == GIT_ENOTFOUND)
-				continue;
-
-			/* treat the file as unreadable */
-			memset(&statbuf, 0, sizeof(statbuf));
-			statbuf.st_mode = GIT_FILEMODE_UNREADABLE;
-
-			error = 0;
-		}
-
-		iter->base.stat_calls++;
+		if (statbuf.st_mode == 0 &&
+		    diriter.d_type == DT_DIR &&
```

Comment:
This optimization avoids calling `stat()` on directories. About 9% fewer `stat()` calls in chromium.